core/network/warp/protocol.hpp

```diff
@@ -47,7 +53,75 @@ namespace kagome::network {
 
     void onTxRequest(const RequestType &) override {}
 
+    /**
+     * @brief Sends a request to a randomly selected peer.
+     *
+     * This method selects a random peer that supports the necessary protocols
+     * and sends a `WarpRequest` to it. If no suitable peer is found, it tries
+     * to use any connected peer.
+     *
+     * @param req The WarpRequest to be sent.
+     * @param cb The callback function to handle the response.
+     */
+    void random(WarpRequest req, Cb cb) override {
+      // Vector to store the peers that support the required protocols
+      std::vector<PeerId> peers;
+
+      // Get the list of protocol IDs
+      auto &protocols1 = base().protocolIds();
+      // Convert protocol IDs to a set for easier comparison
+      std::set protocols2(protocols1.begin(), protocols1.end());
+
+      // Get references to the protocol repository and connection manager
+      auto &protocol_repo =
+          base().host().getPeerRepository().getProtocolRepository();
+      auto &conns = base().host().getNetwork().getConnectionManager();
+
+      // Loop through all peers in the protocol repository
+      for (auto &peer : protocol_repo.getPeers()) {
+        // Skip peers without an active connection
+        if (not conns.getBestConnectionForPeer(peer)) {
+          continue;
+        }
+
+        // Check if the peer supports the required protocols
+        auto common = protocol_repo.supportsProtocols(peer, protocols2);
+        if (not common) {
+          continue;
+        }
+
+        // Skip peers that don't support any of the required protocols
+        if (common.value().empty()) {
+          continue;
+        }
+
+        // Add the peer to the list of suitable peers
+        peers.emplace_back(peer);
+      }
+
+      // If no suitable peers were found, add all connected peers to the list
+      if (peers.empty()) {
```

Comment:
PeerIdS of connected peerS should be contained in peer repository, isn't it? This addition looks redundant.

Reply:
First loop iterates `ProtocolRepository` and filters peers which have both connection and protocol.
First loop may return zero peers.
Second loop iterates `ConnectionManager` to get all connected peers.